### DISCUSSION TITLE
Lesson 9 Raffle.sol compiler issue - Please support

### DISCUSSION THREAD

**Discussion by 0x-kalel** [Cyfrin/foundry-full-course-cu#1269] [Q&A]
Hi All,

I am stuck on lesson 9. I understand the code and the raffle, but can not pass the compiler fail issue:

Error Msg:
Error: 
Compiler run failed:
Error (3656): Contract "Raffle" should be marked as abstract.
  --> src/Raffle.sol:37:1:
   |
37 | contract Raffle is VRFConsumerBaseV2 {
   | ^ (Relevant source part starts here and spans across multiple lines).
Note: Missing implementation: 
   --> lib/chainlink-brownie-contracts/contracts/src/v0.8/VRFConsumerBaseV2.sol:122:5:
    |
122 |     function fulfillRandomWords(uint256 requestId, uint256[] memory randomWords) internal virtual;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

I attempted it again with the original raffle.sol code from Patrick collins in a Raffle2.sol file and still got the same error. If I am right the error is from the import file?? but changed it several times. 

any support is much appreciated.
Kind Regards,

**Comment by ghost** [comment]
Can you show me the `fulfillRandomWords` function in your raffle?

**Comment by 0x-kalel** [comment]
Hi Paceliv,

Thanks here is the full code and the error was in the previous msg:

// SPDX-License-Identifier: MIT

// Layout of Contract:
// version
// imports
// errors
// interfaces, libraries, contracts
// Type declarations
// State variables
// Events
// Modifiers
// Functions

// Layout of Functions:
// constructor
// receive function (if exists)
// fallback function (if exists)
// external
// public
// internal
// private
// internal & private view & pure functions
// external & public view & pure functions

pragma solidity ^0.8.19;

import {VRFCoordinatorV2Interface} from "@chainlink/contracts/src/v0.8/interfaces/VRFCoordinatorV2Interface.sol";
import {VRFConsumerBaseV2} from "@chainlink/contracts/src/v0.8/VRFConsumerBaseV2.sol";

contract Raffle is VRFConsumerBaseV2 {
    error Raffle__NotEnoughEthSent();
    error Raffle__TransferFailed();
    error Raffle__RaffleNotOpen();
    error Raffle__UpKeepNotNeeded(
        uint256 currentBalance,
        uint256 numPlayers,
        uint256 raffleState
    );

    // variable for entrance fee, immutable only input once.

    /*Type Declarations*/

    enum RaffleState {
        OPEN, //0
        CALCULATING // 1 AND IF WE HAD MORE STATES THEN 2 ,3 ETC
    }
    /*state variables*/

    uint16 private constant REQUEST_CONFIRMATIONS = 3;
    uint32 private constant NUM_WORDS = 1;

    //@dev duration of the lottery in seconds
    uint256 private immutable i_entranceFee;
    uint256 private immutable i_interval;
    VRFCoordinatorV2Interface private immutable i_vrfCoordinator;
    bytes32 private immutable i_gasLane;
    uint64 private immutable i_subscriptionId;
    uint32 private immutable i_callbackGasLimit;

    address payable[] private s_players;
    uint256 private s_lastTimeStamp;
    address private s_recentWinner;
    RaffleState private s_raffleState;

    //Events triggered
    event EnteredRaffle(address indexed player);
    event PickedWinner(address indexed winner);

    //constructor with entrance fee
    constructor(
        uint256 entranceFee,
        uint256 interval,
        address vrfCoordinator,
        bytes32 gasLane,
        uint64 subscriptionId,
        uint32 callbackGasLimit
    ) VRFConsumerBaseV2(vrfCoordinator) {
        {
            i_entranceFee = entranceFee;
            i_interval = interval;
            i_vrfCoordinator = VRFCoordinatorV2Interface(vrfCoordinator);
            i_gasLane = gasLane;
            i_subscriptionId = subscriptionId;
            i_callbackGasLimit = callbackGasLimit;

            s_raffleState = RaffleState.OPEN;
            s_lastTimeStamp = block.timestamp;
        }
    }

    function enterRaffle() external payable {
        //require(msg.value >= i_entranceFee, "Not enough ETH sent!");
        if (msg.value < i_entranceFee) {
            revert Raffle__NotEnoughEthSent();
        }

        if (s_raffleState != RaffleState.OPEN) {
            revert Raffle__RaffleNotOpen();
        }

        s_players.push(payable(msg.sender));
        // 1. Makes migration easier
        // 2. Makes front end "indexing" easier
        emit EnteredRaffle(msg.sender);
    }

    // When is the winner suppose to be picked?
    /**
     * @dev This is the function that the chainlink Automation nodes call
     * to see if it's time to perform upkeep.
     * The following should be true for this to return true:
     * 1. The time intervanl has passed bewteen raffle runs.
     * 2. The raffle is in the OPEN state.
     * 3. The contract has ETH (aka Players)
     * 4. (Implicit) the subcription is funded with LINK
     */
    function checkUpkeep(
        bytes memory /*Check data*/
    ) public view returns (bool upkeepNeeded, bytes memory /* performData */) {
        bool timeHasPassed = (block.timestamp - s_lastTimeStamp) >= i_interval;
        bool isOpen = RaffleState.OPEN == s_raffleState;
        bool hasBalance = address(this).balance > 0;
        bool hasPlayers = s_players.length > 0;
        upkeepNeeded = (timeHasPassed && isOpen && hasBalance && hasPlayers);
        return (upkeepNeeded, "0x0");
    }

    // Get a random number
    // Use the random number to pick a player
    // Automatically select after some time interval has passed.
    function performUpkeep(bytes calldata /* performData */) external override {
        (bool upkeepNeeded, ) = checkUpkeep("");
        if (!upkeepNeeded) {
            revert Raffle__UpKeepNotNeeded(
                address(this).balance,
                s_players.length,
                uint256(s_raffleState)
            );
        }
        //1 Use Chainlink VRF to get a random number follow these steps:
        //2. Request a RNG
        //3. Get a random number from the RNG
        //4. Use the random number to pick a player
        s_raffleState = RaffleState.CALCULATING;
        i_vrfCoordinator.requestRandomWords(
            i_gasLane,
            i_subscriptionId,
            REQUEST_CONFIRMATIONS,
            i_callbackGasLimit,
            NUM_WORDS
        );
    }

    function fulfillRandonWords(
        uint256 /*requestId*/,
        uint256[] memory randomWords
    ) internal override {
        //CEI
        //Checks:
        //Effects:
        uint256 indexOfWinner = randomWords[0] % s_players.length;
        address payable winner = s_players[indexOfWinner];
        s_recentWinner = winner;
        s_raffleState = RaffleState.OPEN;
        s_players = new address payable[](0);
        s_lastTimeStamp = block.timestamp;
        emit PickedWinner(winner);
        //Interactions:
        (bool success, ) = winner.call{value: address(this).balance}("");
        if (!success) {
            revert Raffle__TransferFailed();
        }
    }

    /** Getter Function */

    function getEntranceFee() external view returns (uint256) {
        return i_entranceFee;
    }
}

**Reply by ghost** [reply]
The name `fulfillRandonWords` contains a typo, fix to `fulfillRandomWords`.

**Comment by 0x-kalel** [comment]
Thank you fixed that issue. Still looking for an error on the override:
Error: 
Compiler run failed:
Error (7792): Function has override specified but does not override anything.
   --> src/Raffle.sol:132:71:
    |
132 |     function performUpkeep(bytes calldata /* performData */) external override {
    |                                                                       ^^^^^^^^

ismailbulale@ismails-MacBook-Pro foundry-smart-contract-lottery-f23 % 

about to look back at the original function and see if something is missing.

**Reply by ghost** [reply]
You need to inherit `AutomationCompatibleInterface` to make the contract compatible with Chainlink Automation and then can override `checkUpkeep` and `performUpkeep`.

Check: https://github.com/Cyfrin/foundry-smart-contract-lottery-f23/blob/main/src/Raffle.sol#L28-L35

**Comment by 0x-kalel** [comment]
Thank you,

I will give this a go. Only issue this was the extend of the code that Patrick ran and complied without the AutomationCompatibleInterface, hence I thought It should compile on my side to. Thanks again I will try to see if it works in the end.
